sympy/physics/continuum_mechanics/beam.py

```diff
@@ -70,29 +71,43 @@ def __init__(self, length, elastic_modulus, second_moment, variable=Symbol('x'),
 
         Parameters
         ==========
+
         length : Sympifyable
             A Symbol or value representing the Beam's length.
+
         elastic_modulus : Sympifyable
             A SymPy expression representing the Beam's Modulus of Elasticity.
             It is a measure of the stiffness of the Beam material. It can
             also be a continuous function of position along the beam.
-        second_moment : Sympifyable
-            A SymPy expression representing the Beam's Second moment of area.
-            It is a geometrical property of an area which reflects how its
-            points are distributed with respect to its neutral axis. It can
-            also be a continuous function of position along the beam.
+
+        second_moment : Sympifyable or Geometry object
+            Describes the cross-section of the beam  via a sympy expression
+            representing the Beam's Second moment of area. It is a geometrical
+            property of an area which reflects how its points are distributed
+            with respect to its neutral axis. It can also be a continuous
+            function of position along the beam. Alternatively ``second_moment``
+            can be a geometry object such as a ``Polygon`` from the geometry module
+            representing the shape of the cross-section of the beam. The second moment
+            of area will be computed from the gemetry object internally.
+
         variable : Symbol, optional
             A Symbol object that will be used as the variable along the beam
             while representing the load, shear, moment, slope and deflection
             curve. By default, it is set to ``Symbol('x')``.
+
         base_char : String, optional
             A String that will be used as base character to generate sequential
             symbols for integration constants in cases where boundary conditions
             are not sufficient to solve them.
         """
         self.length = length
         self.elastic_modulus = elastic_modulus
-        self.second_moment = second_moment
+        if isinstance(second_moment, GeometryEntity):
```

Comment:
I just realized that putting this `if` block inside `init` will make it run only once (during the creation of the object). Therefore, making changes to `second_moment` won't change `_cross_section`.
```
>>> b1 = Beam(50, E, Circle((0, 0), r))
>>> b1.cross_section
Circle(Point2D(0, 0), r)
>>> b1.second_moment
pi*r*Abs(r)**3/4
>>> b1.second_moment = Polygon((0, 0), (a, 0), (a, c), (0, c))
>>> b1.cross_section
Circle(Point2D(0, 0), r)   # should be Polygon((0, 0), (a, 0), (a, c), (0, c))
>>> b1.second_moment
Polygon(Point2D(0, 0), Point2D(a, 0), Point2D(a, c), Point2D(0, c))  # should be a*c**3/12

```

Reply:
>  `>>>`b1.second_moment = Polygon((0, 0), (a, 0), (a, c), (0, c))
      b1.cross_section

But do we want this kind of thing to be supported? If so then it would better be with the `cross_section` attribute being changed, like:
``` 
b1.cross_section= Polygon((0, 0), (a, 0), (a, c), (0, c)) 
```
And for this the setter will have to be introduced.

Reply:
And for that we can just add this setter in the current code, making if condition to be in `__init__` where it is ideally supposed to be. But `cross_section` won't be private then
```
@cross_section.setter
    def cross_section(self, s):
        if s:
            self.second_moment = s.second_moment_of_area()[0]
        self._cross_section = s
```

Reply:
> But do we want this kind of thing to be supported?

As long as there is a setter method for `second_moment`, the user should be able to change the property using that setter. 
Saying that we can remove the setter for `second_moment` too and make it read-only but then it won't be consistent with other similar attributes of Beam (`length`, `elastic_modulus` & `variable`).

Reply:
Is there any harm in making the `cross_section` attribute non-private, with a setter?
If there is no harm, then I guess we can go with the setter that I have mentioned above.
It would also help in updating the second moment when the cross-section is updated by the user.

Reply:
If you allow second_moment and cross_section to be mutable you need to make sure a user can add both after creation and that if they add one the other is removed.

Reply:
> make sure a user can't add both simultaneously after creation and that if they add one the other is removed.

I have added tests for it. Please have a look